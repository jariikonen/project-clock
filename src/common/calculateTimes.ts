import { Task } from '../types/ProjectClockData';
import ProjectClockError from './ProjectClockError';

export interface TaskStatus {
  task: string;
  status:
    | 'unstarted'
    | 'started'
    | 'suspended'
    | 'resumed'
    | 'completed'
    | undefined;
  timeSpent: number;
}

const emptyTaskStatus = { task: '', status: undefined, timeSpent: 0 };

function getTaskSubjectStr(task: Task) {
  return task.subject.length > 25
    ? `${task.subject.slice(0, 22)}...`
    : task.subject;
}

function calculateDifference(
  startDate: Date,
  endDate: Date,
  task: Task
): number {
  if (startDate > endDate) {
    throw new ProjectClockError(
      `invalid time period '${startDate.toISOString()}' => '${endDate.toISOString()}' (${getTaskSubjectStr(task)}); start date is later than end date`
    );
  }
  return endDate.getTime() - startDate.getTime();
}

function handleSuspendAndResume(
  begin: string,
  suspend: string[],
  resume: string[] | undefined,
  end: string | undefined,
  task: Task
): TaskStatus {
  const beginDate = new Date(begin);
  const taskStatus: TaskStatus = { ...emptyTaskStatus, task: task.subject };
  if (suspend[0]) {
    taskStatus.status = 'suspended';
    taskStatus.timeSpent = calculateDifference(
      beginDate,
      new Date(suspend[0]),
      task
    );
  }
  if (suspend.length === 1 && resume?.length === 1) {
    if (!end) {
      taskStatus.status = 'resumed';
      taskStatus.timeSpent += calculateDifference(
        new Date(resume[0]),
        new Date(),
        task
      );
    } else {
      taskStatus.status = 'completed';
      taskStatus.timeSpent += calculateDifference(
        new Date(resume[0]),
        new Date(end),
        task
      );
    }
  } else if (suspend.length > 1 && resume) {
    resume.forEach((resumeDateStr, i) => {
      if (suspend[i + 1]) {
        taskStatus.status = 'suspended';
        taskStatus.timeSpent += calculateDifference(
          new Date(resumeDateStr),
          new Date(suspend[i + 1]),
          task
        );
      } else if (end) {
        taskStatus.status = 'completed';
        taskStatus.timeSpent += calculateDifference(
          new Date(resumeDateStr),
          new Date(end),
          task
        );
      } else {
        taskStatus.status = 'resumed';
        taskStatus.timeSpent += calculateDifference(
          new Date(resumeDateStr),
          new Date(),
          task
        );
      }
    });
  }
  return taskStatus;
}

function checkTask(task: Task) {
  const { begin, suspend, resume, end } = task;
  if (suspend && !Array.isArray(suspend)) {
    throw new ProjectClockError(
      `invalid task '${getTaskSubjectStr(task)}'; suspend field is not an array`
    );
  }
  if (resume && !Array.isArray(resume)) {
    throw new ProjectClockError(
      `invalid task '${getTaskSubjectStr(task)}'; resume field is not an array`
    );
  }
  if (end && !begin) {
    throw new ProjectClockError(
      `invalid task '${getTaskSubjectStr(task)}'; end date without begin date`
    );
  }
  if (suspend && !begin) {
    throw new ProjectClockError(
      `invalid task '${getTaskSubjectStr(task)}'; suspend date(s) without begin date`
    );
  }
  const resumeNumber = resume ? resume.length : 0;
  if (suspend && end && resumeNumber !== suspend.length) {
    throw new ProjectClockError(
      `invalid task '${getTaskSubjectStr(task)}'; suspend and end without enough resumes`
    );
  }
  if (resume && !begin) {
    throw new ProjectClockError(
      `invalid task '${getTaskSubjectStr(task)}'; resume date(s) without begin date`
    );
  }
  if (resume && !suspend) {
    throw new ProjectClockError(
      `invalid task '${getTaskSubjectStr(task)}'; resume without suspend`
    );
  }
  if (suspend && resume && resume.length > suspend.length) {
    throw new ProjectClockError(
      `invalid task '${getTaskSubjectStr(task)}'; resumed more times than suspended`
    );
  }
  if (suspend && resume) {
    suspend.forEach((suspendDateStr, i) => {
      if (resume[i]) {
        const suspendDate = new Date(suspendDateStr);
        const resumeDate = new Date(resume[i]);
        if (suspendDate.getTime() > resumeDate.getTime()) {
          throw new ProjectClockError(
            `invalid task '${getTaskSubjectStr(task)}'; suspend date (${suspendDate.toISOString()}) is later than resume date (${resumeDate.toISOString()})`
          );
        }
      }
    });
  }
}

/**
 * Calculates the times spent on each task.
 * @param tasks Array of Task objects.
 * @returns An array of TaskStatus objects holding the status and time usage
 *   information for each task.
 */
export default function calculateTimes(tasks: Task[]): TaskStatus[] {
  const statusArray: TaskStatus[] = tasks.map((task) => {
    checkTask(task);
    const { begin, suspend, resume, end } = task;
    if (!begin) {
      return { task: task.subject, status: 'unstarted', timeSpent: 0 };
    }
    if (!suspend && !end) {
      return {
        task: task.subject,
        status: 'started',
        timeSpent: calculateDifference(new Date(begin), new Date(), task),
      };
    }
    if (end && !suspend && !resume) {
      const beginDate = new Date(begin);
      const endDate = new Date(end);
      return {
        task: task.subject,
        status: 'completed',
        timeSpent: calculateDifference(beginDate, endDate, task),
      };
    }
    if (suspend) {
      return handleSuspendAndResume(begin, suspend, resume, end, task);
    }
    throw new Error(
      `internal error; this should not have happened (${getTaskSubjectStr(task)})`
    );
  });

  return statusArray;
}
