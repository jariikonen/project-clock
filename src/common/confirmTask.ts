import confirm from '@inquirer/confirm';

export default async function confirmTask(message: string): Promise<boolean> {
  const answer = await confirm({
    message,
  });
  return answer;
}
