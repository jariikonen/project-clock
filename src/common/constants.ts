// eslint-disable-next-line import/prefer-default-export
export const ERROR_MESSAGE_TIMESHEET_INSPECTION =
  'An error occurred while inspecting the timesheet file';
