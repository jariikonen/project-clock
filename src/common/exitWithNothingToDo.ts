import { outputError, outputSuccess } from './outputFormatting';

/**
 * Outputs message 'Nothing to VERB.' where VERB is the verb given as argument,
 * and does a normal exit (exit value 0).
 */
export default function exitWithNothingToDo(verb: string, error = false) {
  if (error) {
    outputError(`Nothing to ${verb}.`);
    process.exit(1);
  }
  outputSuccess(`Nothing to ${verb}.`);
  process.exit(0);
}
