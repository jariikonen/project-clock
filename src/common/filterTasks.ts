/**
 * The Project Clock `Task` (hereafter just task) can have the following time
 * properties: `begin`, `end`, `suspend` and `resume`. `Begin` and `end` are
 * single timestamps while `suspend` and `resume` are arrays containing one to
 * many timestamps. The task can be seen as a state machine whose state
 * consists of these time properties. This state machine has the following
 * states, i.e., the possible combinations of time properties:
 *
 *    - unstarted (no time properties),
 *    - started (`begin`),
 *    - suspended (`begin`, `suspend`*n, `resume`*n-1; n > 0),
 *    - resumed (`begin`, `suspend`*n, `resume`*n+1; n >= 1) and
 *    - stopped (`begin`, `suspend`*a, `resume`*b, `end`; a = 0...n, b = a).
 *
 * In the beginning, when a task has just been created but not yet started, it
 * is unstarted. A task is started when it gets the `begin` timestamp. A task
 * becomes suspended when it gets one `suspend` timestamp (or the number of
 * `suspend` timestamps is one larger than the number of `resume` timestamps).
 * From suspended state it can move to resumed or stopped state. The task is in
 * resumed state when it gets at least one `resume` timestamp, and when the
 * number of `resume` timestamps is equal to the number of `suspend`
 * timestamps. From resumed and stopped state the task can move back to
 * suspended or to either resumed or stopped state. A started state can also
 * become stopped, without being suspended and resumed before. A stopped state
 * cannot, how ever, become started again. The states can be illustrated with
 * the following state diagram:
 *
 *                                      /-> resumed <-\
 *  unstarted -> started -> suspended <-               |
 *                       \--------------\-> stopped <-/
 *
 * Different states can be divided into the following conceptually useful
 * groups:
 *
 *    - incomplete: unstarted, started, suspended, resumed,
 *    - complete: stopped
 *    - active: started, suspended, resumed,
 *    - suspendable: started, resumed, stopped,
 *    - resumable: suspended, stopped
 *    - stoppable = active
 *
 * The task is complete when it is stopped. All other states are incomplete.
 * An active task is an incomplete task that has been started. Suspendable
 * state group consists of all states from which a task can become suspended.
 * Likewise resumable consists of all the states from which the task can be
 * resumed, and stoppable consists of all the states from which the task can be
 * stopped.
 *
 * Although a task can be moved from suspended to stopped and from stopped to
 * suspended, the statuses in the entries in the timesheet file are treated as
 * if the transfer was always through the resumed state. Stopping a suspended
 * task is done by adding both `resume` and `end` timestamps with the same time
 * value, and suspending a task from stopped state is done by first moving
 * the `end` timestamp to the end of the `suspend` array and after this adding
 * both new`resume` and `suspend` timestamps. Resuming a task from stopped
 * state is done by moving the `end` timestamp to the end of the `suspend`
 * array and adding a new `resume` timestamp.
 */

import { Task } from '../types/ProjectClockData';
import ProjectClockError from './ProjectClockError';

/** Represents the type of a task state. */
export enum TaskStateType {
  Active = 'active',
  Suspendable = 'suspendable',
  Resumable = 'resumable',
  Stoppable = 'stoppable',
  Unstarted = 'unstarted',
  Complete = 'complete',
  Incomplete = 'incomplete',
}

/**
 * Returns the active tasks found in the given task array.
 * @param tasks Array of tasks.
 * @param taskDescriptor A string that is expected to match the task subject.
 */
export function getActiveTasks(tasks: Task[]) {
  return tasks.filter((task) => task.begin && !task.end);
}

/**
 * Returns the active tasks found in the given task array, which
 * match the given task descriptor.
 * @param tasks Array of tasks.
 * @param taskDescriptor A string that is expected to match the task subject.
 */
export function getMatchingActiveTasks(tasks: Task[], taskDescriptor: string) {
  return tasks.filter(
    (task) => task.begin && !task.end && task.subject.match(taskDescriptor)
  );
}

/**
 * Returns the suspendable tasks found from the given task array.
 * @param tasks Array of tasks.
 * @param taskDescriptor A string that is expected to match the task subject.
 */
export function getSuspendableTasks(tasks: Task[]) {
  return tasks.filter((task) => {
    // stopped
    if (task.end) {
      return true;
    }
    // started
    if (task.begin && !task.suspend && !task.end) {
      return true;
    }
    // resumed
    if (
      task.resume &&
      task.resume.length > 0 &&
      !task.end &&
      task.suspend &&
      task.suspend.length === task.resume.length
    ) {
      return true;
    }
    return false;
  });
}

/**
 * Returns the suspendable tasks found in the given task array, which
 * match the given task descriptor.
 * @param tasks Array of tasks.
 * @param taskDescriptor A string that is expected to match the task subject.
 */
export function getMatchingSuspendableTasks(
  tasks: Task[],
  taskDescriptor: string
) {
  return tasks.filter((task) => {
    // stopped
    if (task.end && task.subject.match(taskDescriptor)) {
      return true;
    }
    // started
    if (
      task.begin &&
      !task.suspend &&
      !task.end &&
      task.subject.match(taskDescriptor)
    ) {
      return true;
    }
    // resumed
    if (
      task.resume &&
      task.resume.length > 0 &&
      !task.end &&
      task.suspend &&
      task.suspend.length === task.resume.length &&
      task.subject.match(taskDescriptor)
    ) {
      return true;
    }
    return false;
  });
}

/**
 * Returns the resumbable tasks found in the given task array.
 * @param tasks Array of tasks.
 * @param taskDescriptor A string that is expected to match the task subject.
 */
export function getResumableTasks(tasks: Task[]) {
  return tasks.filter((task) => {
    // stopped
    if (task.end) {
      return true;
    }
    // suspended
    if (
      (!!task.suspend && !task.resume && !task.end) ||
      (!!task.suspend &&
        !!task.resume &&
        task.suspend.length > task.resume?.length &&
        !task.end)
    ) {
      return true;
    }
    return false;
  });
}

/**
 * Returns the resumable tasks found in the given task array, which
 * match the given task descriptor.
 * @param tasks Array of tasks.
 * @param taskDescriptor A string that is expected to match the task subject.
 */
export function getMatchingResumableTasks(
  tasks: Task[],
  taskDescriptor: string
) {
  return tasks.filter((task) => {
    // stopped
    if (task.end && task.subject.match(taskDescriptor)) {
      return true;
    }
    // suspended
    if (
      (!!task.suspend &&
        !task.resume &&
        !task.end &&
        !!task.subject.match(taskDescriptor)) ||
      (!!task.suspend &&
        !!task.resume &&
        task.suspend.length > task.resume?.length &&
        !task.end &&
        !!task.subject.match(taskDescriptor))
    ) {
      return true;
    }
    return false;
  });
}

/**
 * Returns the stoppable/active tasks found in the given task array.
 * @param tasks Array of tasks.
 * @param taskDescriptor A string that is expected to match the task subject.
 */
export function getStoppableTasks(tasks: Task[]) {
  return getActiveTasks(tasks);
}

/**
 * Returns the stoppable/active tasks found in the given task array, which
 * match the given task descriptor.
 * @param tasks Array of tasks.
 * @param taskDescriptor A string that is expected to match the task subject.
 */
export function getMatchingStoppableTasks(
  tasks: Task[],
  taskDescriptor: string
) {
  return getMatchingActiveTasks(tasks, taskDescriptor);
}

/**
 * Filters tasks in the given task array based on the given task state type. If
 * task descriptor is given, the tasks are also filtered based on that and the
 * task state type.
 * @param tasks Array of tasks.
 * @param taskType Task state type based on which the tasks are filtered.
 * @param taskDescriptor A string that is expected to match the task subject.
 */
export function filterTasks(
  tasks: Task[],
  taskType: TaskStateType,
  taskDescriptor = ''
): [Task[], string] {
  switch (taskType) {
    case TaskStateType.Active:
    case TaskStateType.Stoppable:
      return taskDescriptor
        ? [getMatchingActiveTasks(tasks, taskDescriptor), 'matching active']
        : [getActiveTasks(tasks), 'active'];
    case TaskStateType.Suspendable:
      return taskDescriptor
        ? [
            getMatchingSuspendableTasks(tasks, taskDescriptor),
            'matching suspendable',
          ]
        : [getSuspendableTasks(tasks), 'suspendable'];
    case TaskStateType.Resumable:
      return taskDescriptor
        ? [
            getMatchingResumableTasks(tasks, taskDescriptor),
            'matching resumable',
          ]
        : [getResumableTasks(tasks), 'resumable'];
    default:
      throw new ProjectClockError('switch ran out of options');
  }
}
