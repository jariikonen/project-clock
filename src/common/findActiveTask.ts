import { Task } from '../types/ProjectClockData';
import ProjectClockError from './ProjectClockError';

/**
 * Returns active tasks (i.e., started but not stopped tasks).
 * @param tasks Array of Task objects from which to search.
 * @returns Active tasks (i.e., started but not stopped tasks). An empty array
 *    if no active tasks were found.
 * @throws ProjectClockError with a descriptive message ("ERROR: 'begin' is not
 *    a valid timestamp (begin_timestamp)", "ERROR: 'end' is not a valid
 *    timestamp (end_timestamp)", "ERROR: begin date is later than end date
 *    (task_subject)", "ERROR: more than one active tasks" or "ERROR: no active
 *    task found").
 */
export default function findActiveTask(tasks: Task[]) {
  const activeTasks = tasks.filter((task) => task.begin && !task.end);

  if (activeTasks.length > 1) {
    throw new ProjectClockError('ERROR: more than one active task');
  }
  if (activeTasks.length === 0) {
    throw new ProjectClockError('ERROR: no active task found');
  }
  return activeTasks[0];
}
