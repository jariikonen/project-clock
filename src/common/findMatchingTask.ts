import { Task } from '../types/ProjectClockData';
import ProjectClockError from './ProjectClockError';

/**
 * Returns tasks whose 'subject' matches the taskDescriptor parameter.
 * @param tasks Array of Task objects from which to search.
 * @param taskDescriptor RegEx matcher that should match the 'subject' of the
 *    task.
 * @returns Task objects whose 'subject' matches the taskDescriptor parameter.
 *    An empty array if no matches were found.
 * @throws ProjectClockError with a descriptive message ('ERROR: more than one
 *    task matches the descriptor' or 'ERROR: no matching task found').
 */
export default function findMatchingTask(
  tasks: Task[],
  taskDescriptor: string
) {
  const matchingTasks = tasks.filter((task) =>
    task.subject.match(taskDescriptor)
  );
  if (matchingTasks.length > 1) {
    throw new ProjectClockError(
      'ERROR: more than one task matches the descriptor'
    );
  }
  if (matchingTasks.length === 0) {
    throw new ProjectClockError('ERROR: no matching task found');
  }
  return matchingTasks[0];
}
