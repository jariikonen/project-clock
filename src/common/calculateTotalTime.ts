import { TaskStatus } from './calculateTimes';

/**
 * Returns the total amount of time spent on tasks listen in the times array.
 * @param times Array of TaskStatus objects that contain task status and time
 *    usage information.
 * @returns
 */
export default function calculateTotalTime(times: TaskStatus[]) {
  return times.reduce((accum, task) => accum + task.timeSpent, 0);
}
