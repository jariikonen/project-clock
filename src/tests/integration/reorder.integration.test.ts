import prettyAnsi from 'pretty-ansi';
import { PROJECT_NAME, ROOT_DIR } from '../common/constants';
import { createTestDir, removeTestDir } from '../common/testDirectory';
import { createTestFile } from '../common/testFile';
import {
  Command,
  emptyTimesheetFile,
  faultyTask,
  forceStopped,
  moreThanOneTimesheetFile,
  noPermission,
  noTimesheetFile,
} from '../common/userFriendlyErrorMessages';
import { getTestPaths } from '../common/testPaths';
import execute, {
  DOWN,
  SHIFT_AND_DOWN,
  SHIFT_AND_UP,
  SPACE,
  UP,
} from '../common/childProcessExecutor';

const testSuiteName = 'reorder';
const { testDirName, testDirPath, subdirPath, testFilePath } =
  getTestPaths(testSuiteName);

beforeAll(() => {
  createTestDir(testDirPath);
});

afterAll(() => {
  removeTestDir(testDirPath);
});

beforeEach(() => {
  createTestDir(subdirPath);
});

afterEach(() => {
  removeTestDir(subdirPath);
});

describe('User friendly error messages', () => {
  test('"Reorder" command reports timesheet file errors in a user friendly manner; no timesheet file', () => {
    expect.hasAssertions();
    noTimesheetFile(testDirName, Command.Reorder);
  });

  test('"Reorder" command reports timesheet file errors in a user friendly manner; no permission', () => {
    expect.hasAssertions();
    noPermission(testDirName, Command.Reorder);
  });

  test('"Reorder" command reports timesheet file errors in a user friendly manner; more than one timesheet file', () => {
    expect.hasAssertions();
    moreThanOneTimesheetFile(testDirName, Command.Reorder);
  });

  test('"Reorder" command reports timesheet file errors in a user friendly manner; faulty task', async () => {
    expect.hasAssertions();
    await faultyTask(testDirName, Command.Reorder);
  });

  test('"Reorder" command gives a user friendly error message when the command is force stopped with CTRL+C', async () => {
    expect.hasAssertions();
    await forceStopped(testDirName, Command.Reorder, {
      projectName: PROJECT_NAME,
      tasks: [
        {
          subject: 'First completed task',
          begin: '2024-01-01T00:00:00.000Z',
          end: '2024-01-01T01:00:00.000Z',
        },
        {
          subject: 'Second completed task',
          begin: '2024-01-01T00:00:00.000Z',
          suspend: ['2024-01-01T01:00:00.000Z'],
          resume: ['2024-01-01T02:00:00.000Z'],
          end: '2024-01-01T03:00:00.000Z',
        },
      ],
    });
  });

  test('"Reorder" command reports timesheet file errors in a user friendly manner; timesheet file is empty', async () => {
    expect.hasAssertions();
    emptyTimesheetFile(testDirName, Command.Reorder);
  });
});

describe('"Reorder" command', () => {
  beforeEach(() => {
    const twoHoursAgo = new Date(
      new Date().getTime() - 120 * 60 * 1000
    ).toISOString();

    createTestFile(
      {
        projectName: PROJECT_NAME,
        projectSettings: {
          timeParams: {
            day: 8,
            week: 5,
            month: 20,
            year: 260,
          },
        },
        tasks: [
          {
            subject: 'First completed task',
            begin: '2024-01-01T00:00:00.000Z',
            end: '2024-01-01T01:00:00.000Z',
          },
          {
            subject: 'Second completed task',
            begin: '2024-01-01T00:00:00.000Z',
            suspend: ['2024-01-01T01:00:00.000Z'],
            resume: ['2024-01-01T02:00:00.000Z'],
            end: '2024-01-01T03:00:00.000Z',
          },
          {
            subject: 'First active task',
            begin: '2024-01-01T01:00:00.000Z',
            suspend: ['2024-01-01T02:00:00.000Z'],
          },
          {
            subject: 'Second active task',
            begin: '2024-01-01T02:00:00.000Z',
            suspend: ['2024-01-01T03:00:00.000Z', '2024-01-01T04:30:00.000Z'],
            resume: ['2024-01-01T04:00:00.000Z'],
          },
          {
            subject: 'Third active task',
            begin: twoHoursAgo,
          },
          {
            subject: 'First incomplete but not active task',
          },
        ],
      },
      testFilePath
    );
  });

  test('displays correct items', async () => {
    expect.hasAssertions();
    const response = await execute(
      `cd ${subdirPath} && node ${ROOT_DIR}/bin/pclock.js reorder`,
      ['\n'],
      { ...process.env, FORCE_COLOR: '0' }
    );

    expect(prettyAnsi(response)).toMatchSnapshot();
  });

  test('reorders tasks correctly', async () => {
    expect.hasAssertions();
    const response = await execute(
      `cd ${subdirPath} && node ${ROOT_DIR}/bin/pclock.js reorder`,
      [
        SHIFT_AND_DOWN,
        DOWN,
        DOWN,
        SHIFT_AND_UP,
        UP,
        SPACE,
        UP,
        SPACE,
        'b',
        'm',
        '\n',
      ],
      { ...process.env, FORCE_COLOR: '0' }
    );

    expect(prettyAnsi(response)).toMatchSnapshot();
  });
});
