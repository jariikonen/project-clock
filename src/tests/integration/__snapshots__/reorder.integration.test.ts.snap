// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"Reorder" command displays correct items 1`] = `
"? Arrange items of the test project project. (<?> help, <space> select, <a>
toggle all, <i> invert selection, <ctrl/shift/meta + up/j, down/k, pgup/t or
pgdown/b> move selected items, <ctrl/shift/meta + left/h or right/l> move single
 items, <m or M> move selected above or below cursor, <enter> proceed)
   Task                                                 Time       Status
❯◯ First completed task                                 1h         completed
 ◯ Second completed task                                2h         completed
 ◯ First active task                                    1h         suspended
 ◯ Second active task                                   1h 30min   suspended
 ◯ Third active task                                    2h         started
 ◯ First incomplete but not active task                 -          unstarted
<hideCursor>
<moveCursorToColumn77>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorToColumn1>
✔ Arrange items of the test project project.
<moveCursorToColumn45>
<showCursor>
"
`;

exports[`"Reorder" command reorders tasks correctly 1`] = `
"? Arrange items of the test project project. (<?> help, <space> select, <a>
toggle all, <i> invert selection, <ctrl/shift/meta + up/j, down/k, pgup/t or
pgdown/b> move selected items, <ctrl/shift/meta + left/h or right/l> move single
 items, <m or M> move selected above or below cursor, <enter> proceed)
   Task                                                 Time       Status
❯◯ First completed task                                 1h         completed
 ◯ Second completed task                                2h         completed
 ◯ First active task                                    1h         suspended
 ◯ Second active task                                   1h 30min   suspended
 ◯ Third active task                                    2h         started
 ◯ First incomplete but not active task                 -          unstarted
<hideCursor>
<moveCursorToColumn77>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorToColumn1>
? Arrange items of the test project project.
   Task                                                 Time       Status
 ◯ Second completed task                                2h         completed
❯◯ First completed task                                 1h         completed
 ◯ First active task                                    1h         suspended
 ◯ Second active task                                   1h 30min   suspended
 ◯ Third active task                                    2h         started
 ◯ First incomplete but not active task                 -          unstarted
<hideCursor>
<moveCursorToColumn77>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorToColumn1>
? Arrange items of the test project project.
   Task                                                 Time       Status
 ◯ Second completed task                                2h         completed
 ◯ First completed task                                 1h         completed
❯◯ First active task                                    1h         suspended
 ◯ Second active task                                   1h 30min   suspended
 ◯ Third active task                                    2h         started
 ◯ First incomplete but not active task                 -          unstarted
<hideCursor>
<moveCursorToColumn77>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorToColumn1>
? Arrange items of the test project project.
   Task                                                 Time       Status
 ◯ Second completed task                                2h         completed
 ◯ First completed task                                 1h         completed
 ◯ First active task                                    1h         suspended
❯◯ Second active task                                   1h 30min   suspended
 ◯ Third active task                                    2h         started
 ◯ First incomplete but not active task                 -          unstarted
<hideCursor>
<moveCursorToColumn77>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorToColumn1>
? Arrange items of the test project project.
   Task                                                 Time       Status
 ◯ Second completed task                                2h         completed
 ◯ First completed task                                 1h         completed
❯◯ Second active task                                   1h 30min   suspended
 ◯ First active task                                    1h         suspended
 ◯ Third active task                                    2h         started
 ◯ First incomplete but not active task                 -          unstarted
<hideCursor>
<moveCursorToColumn77>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorToColumn1>
? Arrange items of the test project project.
   Task                                                 Time       Status
 ◯ Second completed task                                2h         completed
❯◯ First completed task                                 1h         completed
 ◯ Second active task                                   1h 30min   suspended
 ◯ First active task                                    1h         suspended
 ◯ Third active task                                    2h         started
 ◯ First incomplete but not active task                 -          unstarted
<hideCursor>
<moveCursorToColumn77>
<moveCursorToColumn78>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorToColumn1>
? Arrange items of the test project project.
   Task                                                 Time       Status
 ◯ Second completed task                                2h         completed
❯◉ First completed task                                 1h         completed
 ◯ Second active task                                   1h 30min   suspended
 ◯ First active task                                    1h         suspended
 ◯ Third active task                                    2h         started
 ◯ First incomplete but not active task                 -          unstarted
<hideCursor>
<moveCursorToColumn77>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorToColumn1>
? Arrange items of the test project project.
   Task                                                 Time       Status
❯◯ Second completed task                                2h         completed
 ◉ First completed task                                 1h         completed
 ◯ Second active task                                   1h 30min   suspended
 ◯ First active task                                    1h         suspended
 ◯ Third active task                                    2h         started
 ◯ First incomplete but not active task                 -          unstarted
<hideCursor>
<moveCursorToColumn77>
<moveCursorToColumn78>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorToColumn1>
? Arrange items of the test project project.
   Task                                                 Time       Status
❯◉ Second completed task                                2h         completed
 ◉ First completed task                                 1h         completed
 ◯ Second active task                                   1h 30min   suspended
 ◯ First active task                                    1h         suspended
 ◯ Third active task                                    2h         started
 ◯ First incomplete but not active task                 -          unstarted
<hideCursor>
<moveCursorToColumn77>
<moveCursorToColumn78>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorToColumn1>
? Arrange items of the test project project.
   Task                                                 Time       Status
 ◉ Second completed task                                2h         completed
 ◉ First completed task                                 1h         completed
 ◯ Second active task                                   1h 30min   suspended
 ◯ First active task                                    1h         suspended
 ◯ Third active task                                    2h         started
❯◯ First incomplete but not active task                 -          unstarted
<hideCursor>
<moveCursorToColumn77>
<moveCursorToColumn78>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorToColumn1>
? Arrange items of the test project project.
   Task                                                 Time       Status
 ◯ Second active task                                   1h 30min   suspended
 ◯ First active task                                    1h         suspended
 ◯ Third active task                                    2h         started
 ◉ Second completed task                                2h         completed
 ◉ First completed task                                 1h         completed
❯◯ First incomplete but not active task                 -          unstarted
<hideCursor>
<moveCursorToColumn77>
<moveCursorToColumn73>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorUpBy1Row>
<eraseLine>
<moveCursorToColumn1>
✔ Arrange items of the test project project.
<moveCursorToColumn45>
<showCursor>
"
`;
